language/move-prover/bytecode/src/borrow_analysis.rs

```diff
@@ -254,90 +203,170 @@ impl BorrowInfo {
     }
 
     fn add_edge(&mut self, parent: BorrowNode, child: BorrowNode, weight: BorrowEdge) -> bool {
-        if self.unchecked_nodes.contains(&parent) {
-            // If the parent node is unchecked, so is the child node.
-            self.unchecked_nodes.insert(child.clone());
-        }
-        match self
-            .borrowed_by
-            .entry(parent.clone())
-            .or_default()
-            .entry(child.clone())
+        self.borrowed_by
+            .entry(parent)
             .or_default()
-        {
-            EdgeDomain::Top => false,
-            EdgeDomain::EdgeSet(edges) => match weight {
-                BorrowEdge::Weak => {
-                    self.borrowed_by
-                        .entry(parent)
-                        .or_default()
-                        .insert(child, EdgeDomain::Top);
-                    true
-                }
-                BorrowEdge::Strong(edge) => edges.insert(edge),
-            },
-        }
+            .insert((child, weight))
     }
 
     fn consolidate(&mut self) {
-        for (src, outgoing) in &self.borrowed_by {
-            for (dst, edges) in outgoing {
+        for (src, outgoing) in self.borrowed_by.iter() {
+            for (dst, edge) in outgoing.iter() {
                 self.borrows_from
                     .entry(dst.clone())
                     .or_default()
-                    .insert(src.clone(), edges.clone());
+                    .insert((src.clone(), edge.clone()));
+            }
+        }
+    }
+
+    /// Collect those leaves which are returned and summarize them in a hyper edge.
+    /// Each of those leaves has a path `out_mut -> ref1 .. -> refn -> in_mut`.
+    /// We create a hyper edge `out_mut --summarize(ref1, .., refn)-> in_mut` for it.
+    fn summarize(
```

Comment:
I don't understand how the summarization is happening.  It seems like something akin to "existential quantification" of all edges incident to or from local variables of the function must be happening.  So, for example, suppose there is an input parameter x, a local variable l, an output parameter y, where all are mutable variables.  Suppose there is an edge e from x to l and another edge f from l to y, then the summary should contain [e,f].  In fact, we have to consider all such combinations to generate all possible vectors of edges.

Reply:
We are construction all paths from in to out. This is happening in function construct_hyper_edges. As you suspect these are then all combinations. So a graph with 2 x 2 branches will end up with 4 hyper edges after summarization.